pkg/policy/manager.go

```diff
@@ -0,0 +1,455 @@
+
+	// DownloadPolicy downloads the given policy.
+	DownloadPolicy(ctx context.Context, p *Policy) error
+
+	// RunPolicy runs the given policy.
+	RunPolicy(ctx context.Context, execRequest *ExecuteRequest) (*ExecutionResult, error)
+}
+
+// NewManager returns the manager instance.
+func NewManager(policyDir string, pool *pgxpool.Pool) Manager {
+	// Singleton instantiation
+	once.Do(func() {
+		managerInstance = &ManagerImpl{
+			policyDirectory: policyDir,
+			pool:            pool,
+		}
+	})
+	return managerInstance
+}
+
+// ParsePolicyHubPath parses and validates the given arguments into the Policy struct.
+// Given args should follow the following semantic structure:
+// [(organization/)repository-name(@tag)] ([repository-path])
+func (m *ManagerImpl) ParsePolicyHubPath(args []string, subPolicyPath string) (*Policy, error) {
+	// Make sure the mandatory args are given
+	if len(args) < 1 {
+		return nil, fmt.Errorf("invalid policy path. Repository name is required but got %#v", args)
+	}
+	policy := &Policy{
+		SubPath: subPolicyPath,
+	}
+
+	// Parse and validate org/repository
+	orgRepoSplit := strings.Split(args[0], pathDelimiter)
+
+	// Parse org/repo
+	switch len(orgRepoSplit) {
+	case 2:
+		policy.Organization = orgRepoSplit[0]
+		policy.Repository = orgRepoSplit[1]
+	case 1:
+		policy.Repository = orgRepoSplit[0]
+		policy.Organization = cloudQueryOrg
+	default:
+		return nil, fmt.Errorf("invalid policy path. Repository name malformed: %s", args[0])
+	}
+
+	// Parse version
+	versionSplit := strings.Split(policy.Repository, versionDelimiter)
+	if len(versionSplit) == 2 {
+		policy.Version = versionSplit[1]
+		policy.Repository = versionSplit[0]
+	}
+
+	// Parse repository path if given
+	if len(args) == 2 {
+		policy.RepositoryPath = args[1]
+	}
+	return policy, nil
+}
+
+// DownloadPolicy downloads the given policy from GitHub and stores it in the local policy directory.
+func (m *ManagerImpl) DownloadPolicy(ctx context.Context, p *Policy) error {
+	// Make sure that the local policy organization folder exists
+	osFs := file.NewOsFs()
+	policyOrgFolder := filepath.Join(m.policyDirectory, defaultLocalSubPath, p.Organization)
+	if err := osFs.MkdirAll(policyOrgFolder, 0744); err != nil {
+		return fmt.Errorf("failed to create organization policy directory: %s", policyOrgFolder)
+	}
+
+	// Get GitHub URL
+	gitURL, err := p.getGitHubURL()
+	if err != nil {
+		return fmt.Errorf("failed to parse GitHub URL: %s", err.Error())
+	}
+
+	// Define clone options
+	cloneOptions := &git.CloneOptions{
+		URL:  gitURL,
+		Tags: git.AllTags,
+	}
+
+	// Output progress information if necessary
+	if ui.IsTerminal() {
+		// Print initial information
+		switch {
+		case p.Version != "":
+			ui.ColorizedOutput(ui.ColorProgress, fmt.Sprintf("Cloning Policy %s/%s@%s\n", p.Organization, p.Repository, p.Version))
+		default:
+			ui.ColorizedOutput(ui.ColorProgress, fmt.Sprintf("Cloning Policy %s/%s\n", p.Organization, p.Repository))
+		}
```

Comment:
this can be set outside of the condition since it's checked inside, maybe we should upgrade ui.ColorizedOutput to print to global log if ui isn't terminal.

Reply:
Good point. Yeah that sounds reasonable.